### PR TITLE
Update NodeJS Template Version to 12

### LINKED ISSUE #11650: The generator project - Deprecate node v8 versions
##### **Overview of the feature request**

Node v8 is not longer supported, and node v10 is in maintenance mode (v12 being the LTS version).

##### **Motivation for or Use Case**

First motivation: we can use the newer features of node (`mkdir('p', ...)`), bump (I'm talking semver-major updates) some dependencies.
Second motivation: JCore is gonna do the same.
Third motivation: remove the `mkdirp` dependency, because it won't be needed anymore.

##### **Related issues or PR**

Didn't find any.

-   [X] Yes, I do solemnly swear that I've read everything.

### PR DESCRIPTION
**Pull request by SudharakaP**
This updates NodeJS in templates to version 12.

Related to #11650

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by MathieuAA** [comment]
I think this can be merged unless a build fails

File: generators/server/templates/package.json.ejs [hunk 1]
```diff
@@ -33,6 +33,6 @@
     "@openapitools/openapi-generator-cli": "1.0.12-4.3.0"
   },
   "engines": {
-    "node": ">=8.9.0"
+    "node": "12.16.1"
```
Comment (mshima) [inline generators/server/templates/package.json.ejs:36]: Do you mean to force a specific version?

**Comment by mshima** [comment]
Is there a reason to don’t support node 10 too?

File: generators/server/templates/package.json.ejs (hunk 1 above)
Comment (SudharakaP) [reply generators/server/templates/package.json.ejs:36]: @machima: Not really, that was a mistake. Thanks. :+1:

File: generators/client/templates/react/package.json.ejs [hunk 2]
```diff
@@ -186,7 +186,7 @@ limitations under the License.
   },
 <%_ } _%>
   "engines": {
-    "node": ">=8.9.0"<% if (clientPackageManager === 'yarn') { %>,
+    "node": ">=12.6.1"<% if (clientPackageManager === 'yarn') { %>,
```
Comment (mshima) [inline generators/client/templates/react/package.json.ejs:189]:
```suggestion
    "node": ">=12.16.1"<% if (clientPackageManager === 'yarn') { %>,
```

**Comment by SudharakaP** [comment]
> Is there a reason to don’t support node 10 too?

I think you are right. We should do Node 10 since it's also Active; https://nodejs.org/en/about/releases/

File: generators/client/templates/react/package.json.ejs (hunk 2 above)
Comment (SudharakaP) [reply generators/client/templates/react/package.json.ejs]: I've updated Node to 10 instead. :smile:

**Comment by mshima** [comment]
https://github.com/jhipster/generator-jhipster/pull/11657 was merged so:
There is no reason to make the generator's node >=12.6.1 and keep the templates generated with node >=12.6.1 requiring node >=10.20.1

You should extract a constant like @murdos did at https://github.com/jhipster/generator-jhipster/pull/11645/files see: https://github.com/jhipster/generator-jhipster/blob/c3b06470e66b5272d9b44f5428bb56b3b33f0532/generators/generator-constants.js#L50

**Comment by SudharakaP** [comment]
@mshima: I was under the impression that we should give users the flexibility of using the actively maintained Node versions; whereas for `generator-jhispter` we can use whatever most recent Node version we prefer.
 
I think with the approach you are describing both (templates + generator) will have the same node version right? :smile:

**Comment by mshima** [comment]
> I think with the approach you are describing both (templates + generator) will have the same node version right? 😄

Yep, same version makes more sense to me. I don't see a reason for them to differ.
But I think we should allow node >=10.20.1, and only move to 12 when 10 is deprecated.

**Comment by pascalgrimaud** [comment]
I'm in favor of merging this.

It will probably work for Node 10 but we don't have any builds for that. And I don't want to add another support.

What we use in our build is Node 12, so we should promote this version.

**Comment by murdos** [comment]
Just regarding prettier version: the case is a bit different since we need to use the same version in the generator and in the generated application, since prettier is used to format code on the fly during generation. Otherwise, if versions are different, running prettier again inside generated application, and just after generation might generates changes and that would be awkward for our users.

**Comment by mshima** [comment]
> Just regarding prettier version: the case is a bit different since we need to use the same version in the generator and in the generated application, since prettier is used to format code on the fly during generation. Otherwise, if versions are different, running prettier again inside generated application, and just after generation might generates changes and that would be awkward for our users.

I agree, prettier should use the exact same version.

A bit but not entirely different, what I am saying is that if we are supporting generating the application on node `>=12.16.1`, we should support running the generated application on `>=12.16.1`.

But since engines is only advisory, there will be only a warning if generator-jhipster is installed on node 10.

**Comment by SudharakaP** [comment]
> A bit but not entirely different, what I am saying is that if we are supporting generating the application on node `>=12.16.1`, we should support running the generated application on `>=12.16.1`.
> 
> But since engines is only advisory, there will be only a warning if generator-jhipster is installed on node 10.

I am okay with both ways, so I've went ahead and updated it back to 12. Also I've used the `NODE_VERSION` variable so that we have the same version everywhere. Let me know if you see any further issues. :smile:

**Comment by vishal423** [comment]
It seems a bit rigid to force users to always have the latest nodejs version to be able to use jhipster cli. I just tried creating a new application from the master, and noticed an error:

```
You are running Node version 12.13.1
JHipster requires Node version >=12.16.1
Please update your version of Node.
```

**Comment by murdos** [comment]
> It seems a bit rigid to force users to always have the latest nodejs version to be able to use jhipster cli. I just tried creating a new application from the master, and noticed an error:
> 
 
I agree. I got the same error. I vote to downgrade to 10 as minimum.

**Comment by SudharakaP** [comment]
> > It seems a bit rigid to force users to always have the latest nodejs version to be able to use jhipster cli. I just tried creating a new application from the master, and noticed an error:
> 
> I agree. I got the same error. I vote to downgrade to 10 as minimum.

I agree. So I've went ahead and created a PR to set the generator and template node version to 10.20.1. :smile:

**Pull request merged**